plutus-core/plutus-ir/test/NamesSpec.hs

```diff
@@ -2,19 +2,69 @@ module NamesSpec
     ( names
     ) where
 
+import PlutusPrelude
 
 import PlutusIR.Generators.AST
 import PlutusIR.Mark
+import PlutusIR.Transform.Beta
+import PlutusIR.Transform.CaseReduce
+import PlutusIR.Transform.CommuteFnWithConst
+import PlutusIR.Transform.DeadCode
+import PlutusIR.Transform.EvaluateBuiltins
+import PlutusIR.Transform.Inline.Inline qualified as Inline
+import PlutusIR.Transform.KnownCon
+import PlutusIR.Transform.LetFloatIn qualified as In
+import PlutusIR.Transform.LetMerge
+import PlutusIR.Transform.NonStrict
+import PlutusIR.Transform.RecSplit
 import PlutusIR.Transform.Rename
+import PlutusIR.Transform.ThunkRecursions
+import PlutusIR.Transform.Unwrap
 
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults
 import PlutusCore.Rename
 import PlutusCore.Test
 
 import Test.Tasty
 
 names :: TestTree
-names =
-    testGroup "names"
-        [ test_scopingGood genProgram rename
-        , test_scopingBad genProgram markNonFreshProgram renameProgramM
-        ]
+names = testGroup "names"
+    [ test_scopingGood "beta-reduction" genTerm BindingRemovalNotOk PrerenameYes $
+        pure . beta
+    , test_scopingGood "case-of-known-constructor" genTerm BindingRemovalNotOk PrerenameYes $
+        pure . caseReduce
+    , test_scopingGood "'commuteDefaultFun'" genTerm BindingRemovalNotOk PrerenameYes $
+        pure . commuteDefaultFun
+    , -- We say that it's fine to remove bindings, because they never actually get removed,
+      -- because the scope checking machinery doesn't create unused bindings, every binding
+      -- gets referenced at some point at least once (usually very close to the binding site).
+      -- So this test doesn't really test much.
+      test_scopingGood "dead code elimination" genTerm BindingRemovalNotOk PrerenameNo $
+        removeDeadBindings def
+    , test_scopingGood "constant folding" genTerm BindingRemovalNotOk PrerenameYes $
+        pure . evaluateBuiltins False def defaultBuiltinCostModel
+    , test_scopingBad "inlining" genTerm BindingRemovalOk PrerenameYes $
+        Inline.inline mempty def
```

Comment:
It's `test_scopingBad`, because the inliner does not preserve global uniqueness as it promises due to the lack of marking (or initial renaming of the entire program, which would perform marking too). We discussed that in Slack, I'll create a Jira ticket for this issue.

Reply:
Damn it, draft PRs Roman, draft them! It's the only way to keep Michael away while you're working on the PR!

(just kidding, although I indeed had addressed your comment before seeing it)

Reply:
I've got to procrastinate somehow!!